File: Embarcadero/Berlin/Object Pascal/Beacons/Beacons Proximity/BeaconsRender.pas
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

unit BeaconsRender;

interface

uses
  System.Types, System.UITypes, System.Generics.Collections, System.Math,
  FMX.Types, FMX.Graphics, FMX.Objects, FMX.TextLayout;

type
  TBeaconGraphicInfo = record
    FName: string;
    FColor: TAlphaColor;
    FOriginalColor: TAlphaColor;
    FDistance: Double;
  end;

  TRenderer = class
  private
    FFillBrush: TBrush;
    FStrokeBrush: TStrokeBrush;
    FReferencePathData, FPathData: TPathData;
    FList: TList<TBeaconGraphicInfo>;
    FMaxDistance: Double;
    FTextLayout: TTextLayout;

    procedure GetExtrems(const ARect: TRectF; var AUnit: Single; var AOrigin, ADestination: TPointF);
  public
    constructor Create;
    destructor Destroy; override;
    procedure Render(MaxDistance: Double; const AList: TList<TBeaconGraphicInfo>; Sender: TRectangle;
      Canvas: TCanvas; const ARect: TRectF);
    function GetObjectUnderMouse(const ARectangle: TRectangle; const APoint: TPointF): string;
  end;

implementation

uses
  System.SysUtils, System.Math.Vectors;

const
  MarkScaleX = 5;
  MarkScaleY = 13;

{ TRenderer }

constructor TRenderer.Create;
begin
  inherited;
  FFillBrush := TBrush.Create(TBrushKind.Solid, TAlphaColorRec.Red);
  FStrokeBrush := TStrokeBrush.Create(TBrushKind.Solid, TAlphaColorRec.Red);

  FReferencePathData := TPathData.Create;
  FReferencePathData.MoveTo(TPointF.Create(0, 0));
  FReferencePathData.LineTo(TPointF.Create(1, 1));
  FReferencePathData.LineTo(TPointF.Create(1, 2));
  FReferencePathData.LineTo(TPointF.Create(-1, 2));
  FReferencePathData.LineTo(TPointF.Create(-1, 1));
  FReferencePathData.ClosePath;

  FPathData := TPathData.Create;

  FList := TList<TBeaconGraphicInfo>.Create;
  FMaxDistance := 100;

  FTextLayout := TTextLayoutManager.DefaultTextLayout.Create;;
end;

destructor TRenderer.Destroy;
begin
  FFillBrush.Free;
  FStrokeBrush.Free;
  FTextLayout.Free;
  FReferencePathData.Free;
  FPathData.Free;
  FList.Free;
  inherited;
end;

procedure TRenderer.GetExtrems(const ARect: TRectF; var AUnit: Single; var AOrigin, ADestination: TPointF);
begin
//  AUnit := Min(ARect.Width, ARect.Height) * 0.01;
  AUnit := ARect.Width * 0.003;

  AOrigin.X := 10 * AUnit;
  AOrigin.Y := (ARect.Top + ARect.Bottom) * 0.5;

  ADestination.X := ARect.Right - AOrigin.X;
  ADestination.Y := AOrigin.Y;
end;

function TRenderer.GetObjectUnderMouse(const ARectangle: TRectangle; const APoint: TPointF): string;
var
  LUnit: Single;
  LOrigin, LDestination: TPointF;
  LVector, LAux: TPointF;
  LTouchRect: TRectF;
  I: Integer;
begin
  Result := '';

  GetExtrems(ARectangle.LocalRect, LUnit, LOrigin, LDestination);

  LVector := LDestination - LOrigin;

  for I := 0 to FList.Count - 1 do
  begin
    LAux := LOrigin + LVector * (FList.List[I].FDistance / FMaxDistance);
    LTouchRect.Left := LAux.X - (LUnit * MarkScaleX);
    LTouchRect.Right := LAux.X + (LUnit * MarkScaleX);
    LTouchRect.Top := LAux.Y;
    LTouchRect.Bottom := LAux.Y + (LUnit * MarkScaleY * 2);

    if LTouchRect.Contains(APoint) then
      Exit(FList.List[I].FName);
  end;

end;


procedure TRenderer.Render(MaxDistance: Double; const AList: TList<TBeaconGraphicInfo>; Sender: TRectangle;
  Canvas: TCanvas; const ARect: TRectF);
var
  LOrigin, LDestination: TPointF;
  LCircleOffset, LVector, LAux: TPointF;
  LUnit: Single;
  LIterationCount: Integer;
  I: Integer;
  LState: TCanvasSaveState;
  LMatrix: TMatrix;
begin
  LState := Canvas.SaveState;

  FMaxDistance := MaxDistance;

  GetExtrems(ARect, LUnit, LOrigin, LDestination);

  LVector := LDestination - LOrigin;

  LCircleOffset := TPointF.Create(LUnit * 5, LUnit * 5);

  FStrokeBrush.Thickness := LUnit * 2;
  Canvas.DrawLine(LOrigin, LDestination, 1.0, FStrokeBrush);

  FStrokeBrush.Thickness := 1;
  LIterationCount := 10;
  for I := 0 to LIterationCount - 1 do
  begin
    LAux := LOrigin + LVector * ((I + 1) / LIterationCount);
    Canvas.DrawLine(LAux + TPointF.Create(0, LUnit * 4), LAux - TPointF.Create(0, LUnit * 4), 1.0, FStrokeBrush);
  end;

  FFillBrush.Color := TAlphaColorRec.Red;
  Canvas.FillEllipse(TRectF.Create(LOrigin - LCircleOffset, LOrigin + LCircleOffset), 1.0, FFillBrush);
  Canvas.FillEllipse(TRectF.Create(LDestination - LCircleOffset, LDestination + LCircleOffset), 1.0, FFillBrush);

  if AList <> nil then
  begin
    FList.Count := AList.Count;
    for I := 0 to FList.Count - 1 do
      FList[I] := AList.List[I];
  end;

  for I := 0 to FList.Count - 1 do
  begin
    LAux := LOrigin + LVector * (FList.List[I].FDistance / FMaxDistance);
    FPathData.Assign(FReferencePathData);
    FPathData.Scale(LUnit * MarkScaleX, LUnit * MarkScaleY);
    FPathData.Translate(LAux);
    FFillBrush.Color := FList.List[I].FColor;
    Canvas.FillPath(FPathData, 0.8, FFillBrush);
  end;

  FTextLayout.Text := FMaxDistance.ToString + ' m.';
  FTextLayout.Font.Size := 20;

  LMatrix := Canvas.Matrix * TMatrix.CreateTranslation(ARect.Right - LUnit * 50, 0);
  Canvas.SetMatrix(LMatrix);
  FTextLayout.RenderLayout(Canvas);

  Canvas.RestoreState(LState);
end;

end.

File: Embarcadero/Berlin/CPP/Beacons/Beacons Proximity/BeaconsRender.cpp
//---------------------------------------------------------------------------

// This software is Copyright (c) 2015 Embarcadero Technologies, Inc.
// You may only use this software if you are an authorized licensee
// of an Embarcadero developer tools product.
// This software is considered a Redistributable as defined under
// the software license agreement that comes with the Embarcadero Products
// and is subject to that software license agreement.

//---------------------------------------------------------------------------

#pragma hdrstop

#include "BeaconsRender.h"
#include <System.SysUtils.hpp>
#include <System.Math.Vectors.hpp>
//---------------------------------------------------------------------------

#define MarkScaleX 5
#define MarkScaleY 13

void __fastcall TRenderer::GetExtrems(const TRectF ARect, Single &AUnit, TPointF &AOrigin, TPointF &ADestination)
{
  AUnit = ARect.Width() * 0.003f;

  AOrigin.X = 10 * AUnit;
  AOrigin.Y = (ARect.Top + ARect.Bottom) * 0.5;

  ADestination.X = ARect.Right - AOrigin.X;
  ADestination.Y = AOrigin.Y;
};


TPointF __fastcall TRenderer::CreatePoint(Single AX, Single AY)
{
  TPointF Result;
  Result.X = AX;
  Result.Y = AY;
  return (Result);
};

TRectF __fastcall TRenderer::CreateRectF(Single Left, Single Top, Single Right, Single Bottom)
{
  TRectF Result;
  Result.Left = Left;
  Result.Top = Top;
  Result.Right = Right;
  Result.Bottom = Bottom;
  return(Result);
}

TMatrix __fastcall TRenderer::MultiplyMatrix(TMatrix AMatrix1, TMatrix AMatrix2)
{
  TMatrix Result;
  Result.m11 = AMatrix1.m11 * AMatrix2.m11 + AMatrix1.m12 * AMatrix2.m21 + AMatrix1.m13 * AMatrix2.m31;
  Result.m12 = AMatrix1.m11 * AMatrix2.m12 + AMatrix1.m12 * AMatrix2.m22 + AMatrix1.m13 * AMatrix2.m32;
  Result.m13 = AMatrix1.m11 * AMatrix2.m13 + AMatrix1.m12 * AMatrix2.m23 + AMatrix1.m13 * AMatrix2.m33;
  Result.m21 = AMatrix1.m21 * AMatrix2.m11 + AMatrix1.m22 * AMatrix2.m21 + AMatrix1.m23 * AMatrix2.m31;
  Result.m22 = AMatrix1.m21 * AMatrix2.m12 + AMatrix1.m22 * AMatrix2.m22 + AMatrix1.m23 * AMatrix2.m32;
  Result.m23 = AMatrix1.m21 * AMatrix2.m13 + AMatrix1.m22 * AMatrix2.m23 + AMatrix1.m23 * AMatrix2.m33;
  Result.m31 = AMatrix1.m31 * AMatrix2.m11 + AMatrix1.m32 * AMatrix2.m21 + AMatrix1.m33 * AMatrix2.m31;
  Result.m32 = AMatrix1.m31 * AMatrix2.m12 + AMatrix1.m32 * AMatrix2.m22 + AMatrix1.m33 * AMatrix2.m32;
  Result.m33 = AMatrix1.m31 * AMatrix2.m13 + AMatrix1.m32 * AMatrix2.m23 + AMatrix1.m33 * AMatrix2.m33;
  return(Result);
}


__fastcall TRenderer::TRenderer()
{
  FFillBrush = new TBrush(TBrushKind::Solid, (TAlphaColor)TAlphaColorRec::Red);
  FStrokeBrush =  new TStrokeBrush(TBrushKind::Solid, (TAlphaColor)TAlphaColorRec::Red);

  FReferencePathData = new TPathData;

  FReferencePathData->MoveTo(CreatePoint(0, 0));
  FReferencePathData->LineTo(CreatePoint(1, 1));
  FReferencePathData->LineTo(CreatePoint(1, 2));
  FReferencePathData->LineTo(CreatePoint(-1, 2));
  FReferencePathData->LineTo(CreatePoint(-1, 1));
  FReferencePathData->ClosePath();

  FPathData = new TPathData;

  FMaxDistance = 100;
};

__fastcall TRenderer::~TRenderer()
{
  FFillBrush->Free();
  FStrokeBrush->Free();
  FReferencePathData->Free();
  FPathData->Free();
  FList.Length = 0;
};

void __fastcall TRenderer::Render(Double MaxDistance, TBeaconGraphicList AList,TRectangle *Sender, TCanvas *Canvas, const TRectF ARect)
{
  TPointF LOrigin;
  TPointF LDestination;
  TPointF LCircleOffset;
  TPointF LVector;
  TPointF LAux;
  Single LUnit;
  Integer LIterationCount;
  int I;
  TCanvasSaveState *LState;
  TMatrix LMatrix;

  LState = Canvas->SaveState();

  FMaxDistance = MaxDistance;

  GetExtrems(ARect, LUnit, LOrigin, LDestination);

  LVector = LDestination - LOrigin;

  LCircleOffset = CreatePoint(LUnit * 5, LUnit * 5);

  FStrokeBrush->Thickness = LUnit * 2;
  Canvas->DrawLine(LOrigin, LDestination, 1.0, FStrokeBrush);

  FStrokeBrush->Thickness = 1;
  LIterationCount = 10;
  for (I = 0; I < LIterationCount; I++)
  {
	LAux.X = LOrigin.X + LVector.X * ((I + 1) / LIterationCount);
	LAux.Y = LOrigin.Y + LVector.Y * ((I + 1) / LIterationCount);
	Canvas->DrawLine(LAux + CreatePoint(0, LUnit * 4), LAux - CreatePoint(0, LUnit * 4), 1.0, FStrokeBrush);
  };

  FFillBrush->Color = (TAlphaColor)TAlphaColorRec::Red;
  Canvas->FillEllipse(CreateRectF(LOrigin.X - LCircleOffset.X, LOrigin.Y - LCircleOffset.Y,
                     LOrigin.X + LCircleOffset.X, LOrigin.Y + LCircleOffset.Y), 1.0, FFillBrush);
  Canvas->FillEllipse(CreateRectF(LDestination.X - LCircleOffset.X, LDestination.Y - LCircleOffset.Y,
					 LDestination.X + LCircleOffset.X, LDestination.Y + LCircleOffset.Y), 1.0, FFillBrush);

  if (AList.Length > 0)
  {
	FList.Length = AList.Length;
	for (I = 0; I < FList.Length; I++)
	{
	  FList[I] = AList[I];
	}
  };

  for (I = 0; I < FList.Length; I++)
  {
	LAux.X = LOrigin.X + LVector.X * (FList[I].FDistance / FMaxDistance);
	LAux.Y = LOrigin.Y + LVector.Y * (FList[I].FDistance / FMaxDistance);
	FPathData->Assign(FReferencePathData);
	FPathData->Scale(LUnit * MarkScaleX, LUnit * MarkScaleY);
	FPathData->Translate(LAux);
	FFillBrush->Color = (TAlphaColor)FList[I].FColor;
	Canvas->FillPath(FPathData, 0.8, FFillBrush);
  };

  LMatrix = MultiplyMatrix(Canvas->Matrix, TMatrix::CreateTranslation(ARect.Right - LUnit * 50, 0));
  Canvas->SetMatrix(LMatrix);
  Canvas->RestoreState(LState);
};

String __fastcall TRenderer::GetObjectUnderMouse(TRectangle *ARectangle, const TPointF APoint)
{
  Single LUnit;
  TPointF LOrigin;
  TPointF LDestination;
  TPointF LVector;
  TPointF LAux;
  TRectF LTouchRect;
  int I;
  String Result;

  Result = "";

  GetExtrems(ARectangle->LocalRect, LUnit, LOrigin, LDestination);

  LVector = LDestination - LOrigin;

  for (I = 0; I < FList.Length; I++)
  {
	LAux.X = LOrigin.X + LVector.X * (FList[I].FDistance / FMaxDistance);
	LAux.Y = LOrigin.Y + LVector.Y * (FList[I].FDistance / FMaxDistance);
	LTouchRect.Left = LAux.X - (LUnit * MarkScaleX);
	LTouchRect.Right = LAux.X + (LUnit * MarkScaleX);
	LTouchRect.Top = LAux.Y;
	LTouchRect.Bottom = LAux.Y + (LUnit * MarkScaleY * 2);

	if (LTouchRect.Contains(APoint))
	{
	  Result = FList[I].FName;
	  break;
	}
  };
  return(Result);
};

#pragma package(smart_init)
